### PR TITLE
Implement Universal Font Shaper

### LINKED ISSUE #217: Implement Universal Font Shaper
Now that #198 has been merged the fundamental infrastructure is there to allow support for Bidi text rendering however we still lack some dedicated shapers. This issue tracks the development of a Universal shaper.

Guide.
https://github.com/n8willis/opentype-shaping-documents/blob/master/opentype-shaping-use.md

HarfBuzz Implementation
https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-shaper-use.cc

FontKit Implementation
https://github.com/foliojs/fontkit/blob/master/src/opentype/shapers/UniversalShaper.js

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/Fonts/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

A Work-In-Progress implementation of the [Universal Shaping Engine (USE)](https://docs.microsoft.com/en-us/typography/script-development/use).  Fixes #217 

There's a fair amount to do here but I have some of the groundwork laid out already. I was hoping to base the work primarily on FontKit but I have no idea how to implement the [state machine](https://github.com/foliojs/dfa).

If anyone from the community is reading this PLEASE HELP OUT. I can't possibly figure this all out on my own ❤️.

Guide.
https://github.com/n8willis/opentype-shaping-documents/blob/master/opentype-shaping-use.md

HarfBuzz Implementation
https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-shaper-use.cc

FontKit Implementation
https://github.com/foliojs/fontkit/blob/master/src/opentype/shapers/UniversalShaper.js

TODO:

- [x] Refactor shapers to allow shaping stages.
- [x] Add Indic shaping data.
- [ ] Add USE specific Extended Indic shaping data.
- [x] Implement state machine.
- [x] Implement category lookups.
- [x] Implement shaper.
- [x] Add tests.

I decided to leave attempting to add USE specific Extended Indic shaping data for now as it breaks the working statemachine. I imagine migrating the generation code to contain all the latest harfbuzz changes will mitigate this but it's not worth the effort for now.

Here's some example rendering of Balinese content. Our output is 100% compliant. 
image

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
OK. 

So I have the beginnings of a plan here....

I need to port the state machine used by JavaScript from here https://github.com/foliojs/dfa

Instead of using Pegjs as the PEG (Parsing Expression Grammar) parser generator I can use [Pegasus](http://otac0n.com/Pegasus/). 

However... The original sample really abuses type safety so I'm having real difficulty with my early experiments.

**Comment by stefannikolei** [comment]
> However... The original sample really abuses type safety so I'm having real difficulty with my early porting experiments.

any should go to hell xD

i was looking for other implementations besides the js and c implementation of harfbuzz.

I found a rust port: 
https://github.com/RazrFalcon/rustybuzz/blob/master/src/complex/universal.rs

perhaps this helps

**Comment by JimBobSquarePants** [comment]
> > However... The original sample really abuses type safety so I'm having real difficulty with my early porting experiments.
> 
> any should go to hell xD
> 
> i was looking for other implementations besides the js and c implementation of harfbuzz.
> 
> I found a rust port: https://github.com/RazrFalcon/rustybuzz/blob/master/src/complex/universal.rs
> 
> perhaps this helps

That rust port has an open issue because it's missing a Ragel compiled state machine, the state machine is the bit I probably need most. 

I've had a look at Ragel before, but I cannot make head no tail of it. Supposedly you can build a C# state machine using it. 

https://github.com/adrian-thurston/ragel

**Comment by JimBobSquarePants** [comment]
First test passes in the state machine tests! Fingers crossed this is a sign of things to come.

**Review by tocsoft — COMMENTED** [review]
General question, should we be commiting *.generated.* files or should we be generating them as part of the build?

File: src/SixLabors.Fonts/GlyphShapingData.cs [hunk 1]
```diff
@@ -101,15 +105,44 @@ public GlyphShapingData(GlyphShapingData data, bool clearFeatures = false)
...
             => FormattableString
             .Invariant($" {this.GlyphId} : {this.CodePoint.ToDebuggerDisplay()} : {CodePoint.GetScriptClass(this.CodePoint)} : {this.Direction} : {this.TextRun.TextAttributes} : {this.LigatureId} : {this.LigatureComponent} : {this.IsDecomposed}");
 
         internal string ToDebuggerDisplay() => this.DebuggerDisplay;
     }
+
+    /// <summary>
+    /// Represents information required for universal shaping.
+    /// </summary>
+    internal class USEInfo
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/GlyphShapingData.cs]: Done

File: src/SixLabors.Fonts/GlyphShapingData.cs [hunk 2]
```diff
@@ -101,15 +105,44 @@ public GlyphShapingData(GlyphShapingData data, bool clearFeatures = false)
...
+
         /// <summary>
         /// Gets or sets a value indicating whether this glyph is the result of a decomposition substitution
         /// </summary>
         public bool IsDecomposed { get; set; }
 
+        /// <summary>
+        /// Gets or sets the universal shaping information.
+        /// </summary>
+        public USEInfo? USEInfo { get; set; }
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/GlyphShapingData.cs]: Done

**Comment by JimBobSquarePants** [comment]
> General question, should we be commiting _.generated._ files or should we be generating them as part of the build?

Good question. It's definitely a manual process just now plus there's a bit of a chicken/egg situation going on with the `UnicodeData` class that references all the generated tries. I'm sure there's smarter ways to do it but I always end up regretting messing around with custom build config.

**Review by tocsoft — APPROVED** [review]

**Pull request merged**